### PR TITLE
Permissive iemgui labels

### LINKED ISSUE #880: weird characters like "[" (as label) hang iemgui properties
It seems that this arrangement causes problems:

a canvas, with a label far outside (to the top-right) but with a really large font size - or is it to do with the use of '[' as a label?
Right-click on the canvas object handle and select properties, pd freezes completely and you cannot do anything, and have to kill Pd.

Pd-0.50.2, Linux, Debian 9, 64 bit, intel
largeFontCanvasBug

[largeFontCanvasBug.pd.zip](https://github.com/pure-data/pure-data/files/4128587/largeFontCanvasBug.pd.zip)

### PR DESCRIPTION
**Pull request by HenriAugusto**
This allows the use of `[` and `]` in iemgui's objects labels.

I've tested the labels and checked if the send/receives are working as expected.

_(Tested on Win 10)_

image

Fixes #880

### CODE REVIEW

**Comment by Ant1r** [comment]
great! with this, you can still open the properties dialog, which was hanging Pd before.

**Comment by porres** [comment]
I tested and it does fix the usage of brackets!

I also see that this fixes https://github.com/pure-data/pure-data/issues/417

**Comment by porres** [comment]
hmmm, I thought the deal was to just put brackets in the tcl code, which would be much simpler than put them in the C code and send them to tcl...

**Comment by HenriAugusto** [comment]
> I also see that this fixes #417

Same for me.

**Comment by HenriAugusto** [comment]
> hmmm, I thought the deal was to just put brackets in the tcl code, which would be much simpler than put them in the C code and send them to tcl...

Indeed. Maybe it would better to add brackets in the `*_properties` function for each iemgui object.

```TCL
static void hslider_properties(t_gobj *z, t_glist *owner)
{
    [...]
    sprintf(buf, "pdtk_iemgui_dialog %%s |hsl| \
            --------dimensions(pix)(pix):-------- %d %d width: %d %d height: \
            -----------output-range:----------- %g left: %g right: %g \
            %d lin log %d %d empty %d \
            {%s} {%s} \
            {%s} %d %d \
            [...]);
    gfxstub_new(&x->x_gui.x_obj.ob_pd, x, buf);
}```

I'll update and force-push,

**Comment by porres** [comment]
> Maybe it would better to add brackets in the *_properties function for each iemgui object.

yeah, I was thinking about doing something like today in my other PR ;) https://github.com/pure-data/pure-data/pull/946 but I hadn't really started checking on the idea or thought about it at all...

**Comment by Ant1r** [comment]
I think I like the present C-side unified solution, better than manually editing the string for every object. But certainly it's a matter of taste...

**Comment by umlaeute** [comment]
i agree with @Ant1r that the current solution is preferable than adding the curlies in all invocations of `<iemgui>_properties`.

this makes it much more robust.

File: src/g_all_guis.c [hunk 1]
```diff
@@ -183,6 +183,31 @@ t_symbol *iemgui_raute2dollar(t_symbol *s)
...
+    int i = 0;
+    if (strlen(s->s_name) >= MAXPDSTRING)
+        return (s);
+    for (s1 = s->s_name, s2 = buf; ; s1++, s2++, i++)
+    {
+        if (i == 0)
+        {
+            *s2 = '{';
+            s2++;
+        }
```
Comment (umlaeute) [inline src/g_all_guis.c:199]: please remove the trailing whitespace.

File: src/g_canvas.h [hunk 2]
```diff
@@ -630,6 +630,7 @@ EXTERN void guiconnect_notarget(t_guiconnect *x, double timedelay);
 /* ------------- IEMGUI routines used in other g_ files ---------------- */
 EXTERN t_symbol *iemgui_raute2dollar(t_symbol *s);
 EXTERN t_symbol *iemgui_dollar2raute(t_symbol *s);
+EXTERN t_symbol *iemgui_put_in_braces(t_symbol *s);
```
Comment (umlaeute) [inline src/g_canvas.h:633]: no, please don't export this new function.
as @Ant1r pointed out in https://github.com/pure-data/pure-data/pull/953#issuecomment-616194527 there's already trouble that the `raute2dollar` and `dollar2raute`  functions are exported. (they should be deprecated and retired asap)

File: src/g_all_guis.c (hunk 1 above)
Comment (umlaeute) [inline src/g_all_guis.c:199]: please remove the trailing whitespace.

(side note, git has a template pre-commit hook that takes care of this; you probably should enable it in your local clone of the repository...)

Comment (HenriAugusto) [reply src/g_all_guis.c]: ok. thanks for the tip

**Comment by porres** [comment]
Just mentioning this also fixes other issues besides being able to use "[". For instance, it is necessary for https://github.com/pure-data/pure-data/pull/953

Moreover, I was detecting other similar issues, if you set a label with some characters allowed by Pd via escaping with backlash, you get a tcl error when opening the properties, but this PR fixes it.

See example of that issue below

79516014-7f396c80-8020-11ea-8c51-d6595f21dae5

**Comment by HenriAugusto** [comment]
> while you are there, please also allow labels that have curlies (while you cannot type curlies in Pd, it's easy enough to create them with `[makefilename %c]`, and Pd supports displaying curlies now.
> use `pdgui_strnescape` for this.

Ok this feels very silly but i can't seem to get it working. 
The `pdgui_strnescape()` adds a `'\\'` char before the braces. But when TCL receive the string it receive the backward slashes . 

image


```C
void iemgui_label(void *x, t_iemgui *iemgui, t_symbol *s)
{
    t_symbol *old;
        /* tb: fix for empty label { */
        if (s == gensym(""))
                s = gensym("empty");
        /* tb } */
    old = iemgui->x_lab;
    iemgui->x_lab_unexpanded = iemgui_raute2dollar(s);
    iemgui->x_lab = canvas_realizedollar(iemgui->x_glist, iemgui->x_lab_unexpanded);

    char lab_escaped[MAXPDSTRING];
    lab_escaped[MAXPDSTRING-1] = 0; //is this line necessary?
    pdgui_strnescape( &lab_escaped[0], MAXPDSTRING, iemgui->x_lab->s_name, strlen(iemgui->x_lab->s_name) );
    if(glist_isvisible(iemgui->x_glist) && iemgui->x_lab != old)
        sys_vgui(".x%lx.c itemconfigure %lxLABEL -text {%s} \n",
                 glist_getcanvas(iemgui->x_glist), x,
                 strcmp(s->s_name, "empty")?lab_escaped:"");
}
```

**PS:** I've clicked re-review changes accidentally

**Comment by umlaeute** [comment]
>  But when TCL receive the string it receive the backward slashes.

there's an unescape proc somewhere in Pd-GUI.

/me looks

ah yes: `::pdtk_text::unescape` (and it requires your label to end with a space (` `) that will be stripped away. don't ask)

**Comment by HenriAugusto** [comment]
Ok that was a little bit trickier than i thought.

Users still won't be able to use backward slashes in the labels but i guess it's okay.

**Comment by umlaeute** [comment]
i've squashed and included this PR in #1353 and am therefore closing this in favour of the cleaned up one.

**Pull request closed without merging**